$OpenBSD: patch-event_c,v 1.15 2016/03/01 11:42:24 dcoppa Exp $

commit a181bee46533f92d573f7a4622391ae58320cda8
Author: Uli Schlachter <psychon@znc.in>
Date:   Wed Feb 17 18:52:35 2016 +0100

Fix unbalance Lua stack usage in event_handle_leavenotify()

Commit 42e00819584c removed a call to lua_pop(L,1) that was still necessary.
This commit adds that call again.

Fixes: https://github.com/awesomeWM/awesome/issues/703
Signed-off-by: Uli Schlachter <psychon@znc.in>

commit e54d911a94697eb5fd5b7d5ac2a72e4d3be6194b
Author: Uli Schlachter <psychon@znc.in>
Date:   Sun Feb 28 13:29:25 2016 +0100

Make client key bindings for e.g. xeyes work again

Instead of focusing the root window, we now create a "focus window" inside of
our frame window. This window is placed so that it is not visible, but we can
grab key bindings on it to simulate the window having the input focus.

Fixes: https://github.com/awesomeWM/awesome/issues/699
Signed-off-by: Uli Schlachter <psychon@znc.in>

--- event.c.orig	Sat Jan 30 14:55:18 2016
+++ event.c	Tue Mar  1 11:24:18 2016
@@ -507,6 +507,7 @@ event_handle_leavenotify(xcb_leave_notify_event_t *ev)
     {
         luaA_object_push(globalconf.L, c);
         luaA_object_emit_signal(globalconf.L, -1, "mouse::leave", 0);
+        lua_pop(globalconf.L, 1);
     }
 
     lua_pushnil(globalconf.L);
@@ -644,7 +645,7 @@ event_handle_key(xcb_key_press_event_t *ev)
         /* get keysym ignoring all modifiers */
         xcb_keysym_t keysym = keyresolv_get_keysym(ev->detail, 0);
         client_t *c;
-        if((c = client_getbywin(ev->event)))
+        if((c = client_getbywin(ev->event)) || (c = client_getbynofocuswin(ev->event)))
         {
             luaA_object_push(globalconf.L, c);
             event_key_callback(ev, &c->keys, -1, 1, &keysym);
@@ -835,6 +836,8 @@ event_handle_mappingnotify(xcb_mapping_notify_event_t 
             client_t *c = *_c;
             xcb_ungrab_key(globalconf.connection, XCB_GRAB_ANY, c->window, XCB_BUTTON_MASK_ANY);
             xwindow_grabkeys(c->window, &c->keys);
+            if (c->nofocus_window)
+                xwindow_grabkeys(c->nofocus_window, &c->keys);
         }
     }
 }
